bundles/org.eclipse.osgi/felix/src/org/apache/felix/resolver/Candidates.java

```diff
@@ -1155,44 +1168,70 @@ public void dump(ResolveContext rc)
-                + " (" + ((wiring != null) ? "RESOLVED)" : "UNRESOLVED)"));
-            List<Requirement> reqs = (wiring != null)
-                ? wiring.getResourceRequirements(null)
+            dumpResource(resource, rc, all, printStream);
+        }
+        printStream.println("=== END CANDIDATE MAP ===");
+    }
+
+    protected void dumpResource(Resource resource, ResolveContext rc, boolean all, PrintStream printStream) {
+        Wiring wiring = rc == null ? null : rc.getWirings().get(resource);
+        List<Requirement> reqs = (wiring != null) ? wiring.getResourceRequirements(null)
                 : resource.getRequirements(null);
-            for (Requirement req : reqs)
-            {
-                CandidateSelector candidates = m_candidateMap.get(req);
-                if ((candidates != null) && (!candidates.isEmpty()))
-                {
-                    System.out.println("    " + req + ": " + candidates);
+        List<Requirement> dreqs = (wiring != null) ? Util.getDynamicRequirements(wiring.getResourceRequirements(null))
+                : Util.getDynamicRequirements(resource.getRequirements(null));
+        boolean hasMulti = hasMulti(reqs);
+        printStream.println("  " + (hasMulti ? "[?]" : "[!]") + Util.getResourceName(resource) + " ("
+                + ((wiring != null) ? "RESOLVED)" : "UNRESOLVED)"));
+        if (all || hasMulti) {
+            printRe(reqs, printStream, all);
+            printRe(dreqs, printStream, all);
+        }
+    }
+
+    private boolean hasMulti(List<Requirement> reqs) {
+        for (Requirement req : reqs) {
+            CandidateSelector candidates = m_candidateMap.get(req);
+            if ((candidates != null) && (!candidates.isEmpty())) {
+                List<Capability> remaining = candidates.getRemainingCandidates();
+                if (remaining.size() > 1) {
+                    return true;
                 }
             }
-            reqs = (wiring != null)
-                ? Util.getDynamicRequirements(wiring.getResourceRequirements(null))
-                : Util.getDynamicRequirements(resource.getRequirements(null));
-            for (Requirement req : reqs)
-            {
-                CandidateSelector candidates = m_candidateMap.get(req);
-                if ((candidates != null) && (!candidates.isEmpty()))
-                {
-                    System.out.println("    " + req + ": " + candidates);
+        }
+        return false;
+    }
+
+    protected int printRe(List<Requirement> reqs, PrintStream printStream, boolean all) {
+        int dup = 0;
+        for (Requirement req : reqs) {
+            CandidateSelector candidates = m_candidateMap.get(req);
+            if ((candidates != null) && (!candidates.isEmpty())) {
+                List<Capability> remaining = candidates.getRemainingCandidates();
+                boolean hasMulti = remaining.size() > 1;
+                if (all || hasMulti) {
+                    dup++;
```

Comment:
Adding a Equinox dependency here for tracing that is hard coded to be disabled should not be done

Reply:
Replaced by Felix Resolver Util